### PR TITLE
Merge dev into ama via merge-ama branch

### PR DESCRIPTION
**Pull request by scottleibrand**

### CODE REVIEW

**Comment by scottleibrand** [comment]
@cjo20 can you look at whether this properly merges all your changes into AMA?

**Comment by scottleibrand** [comment]
@sulkaharo and @jasoncalabrese can you look at this and make sure my merge conflict resolution doesn't break anything you've added to AMA recently?

**Comment by sulkaharo** [comment]
Checking!

**Comment by sulkaharo** [comment]
The --export-defaults and --update-preferences features have gone missing from oref0-get-profile. profile/index.js now has some unneeded code; any value that has a default doesn't need additional code for the inputs to be read in. I'll PR some changes later.

All tests for the new features pass; good sign though.

File: bin/oref0-get-profile.js [hunk 1]
```diff
@@ -42,40 +42,42 @@ function updatePreferences (prefs) {
 
 if (!module.parent) {
     
-    var cwd = process.cwd()
-        
-    var pumpsettings_input = process.argv.slice(2, 3).pop();
-
-    if (pumpsettings_input !== undefined && pumpsettings_input.indexOf('export-defaults') > 0) {
```
Comment (cjo20) [inline bin/oref0-get-profile.js]: Should this if statement (and the next one) be removed or not?

**Comment by cjo20** [comment]
I can't see anything obvious missing from browsing through the code.

File: bin/oref0-get-profile.js [hunk 2]
```diff
@@ -42,40 +42,42 @@ function updatePreferences (prefs) {
...
+    var preferences_input = params._.slice(4, 5).pop()
+    var carbratio_input = params._.slice(5, 6).pop()
+    var temptargets_input = params._.slice(6, 7).pop()
+    var model_input = params.model;
+    if (params._.length > 6)
+    {
+      model_input = params.model ? params.params._.slice(5, 6).pop() : false;
+      var carbratio_input = params._.slice(6, 7).pop()
+      var temptargets_input = params._.slice(7, 8).pop()
+    }
```
Comment (cjo20) [inline bin/oref0-get-profile.js]: I'm having a horrible "That shouldn't work" moment; Shouldn't `params.params._.` be `params._`, and doesn't the `var` in front of the two inputs cause them to be local and go out of scope at the end of the if?

**Comment by scottleibrand** [comment]
Yeah, looks like I mistakenly deleted the export-defaults and update-preferences stuff.  I'm not quite sure how to merge it with the yargs stuff, though.  Ideas and PRs welcome.  Here's the diff:

```
@@ -42,40 +42,42 @@ function updatePreferences (prefs) {

 if (!module.parent) {

-    var cwd = process.cwd()
-
-    var pumpsettings_input = process.argv.slice(2, 3).pop();
-
-    if (pumpsettings_input !== undefined && pumpsettings_input.indexOf('export-defaults') > 0) {
-      exportDefaults();
-      process.exit(0);
-    }
-
-    if (pumpsettings_input !== undefined && pumpsettings_input.indexOf('update-preferences') > 0) {
-      var prefs_input = process.argv.slice(3, 4).pop();
-         if (!prefs_input) { console.error('usage: ', process.argv.slice(0, 2), '<preferences.json>'); process.exit(0); }
-      var prefs = require(cwd + '/' + prefs_input);
-      updatePreferences(prefs);
-      process.exit(0);
-    }
-
+    var argv = require('yargs')
+      .usage("$0 pump_settings.json bg_targets.json insulin_sensitivities.json basal_profile.json [preferences.json] [--model model.json] [<carb_ratios.json>] [<temptargets.json>]")
+      .option('model', {
+        alias: 'm',
+        describe: "Pump model response",
+        default: false
+      })
+      .strict(true)
+      .help('help')
+
+    var params = argv.argv;
+    var pumpsettings_input = params._.slice(0, 1).pop()
     if ([null, '--help', '-h', 'help'].indexOf(pumpsettings_input) > 0) {
       usage( );
       process.exit(0);
     }
```

File: bin/oref0-get-profile.js [hunk 3]
```diff
@@ -42,40 +42,42 @@ function updatePreferences (prefs) {
...
-    var temptargets_input = process.argv.slice(8, 9).pop()
-    
+    var bgtargets_input = params._.slice(1, 2).pop()
+    var isf_input = params._.slice(2, 3).pop()
+    var basalprofile_input = params._.slice(3, 4).pop()
+    var preferences_input = params._.slice(4, 5).pop()
+    var carbratio_input = params._.slice(5, 6).pop()
+    var temptargets_input = params._.slice(6, 7).pop()
+    var model_input = params.model;
+    if (params._.length > 6)
```
Comment (jasoncalabrese) [inline bin/oref0-get-profile.js]: both `carbratio_input` and `temptargets_input` are off by 1 for me using the same ini the I was using for `advanced-meal-assist` with the addition `remainder = --model monitor/model.json`

to make it work I needed:
```
carbratio_input = params._.slice(5, 6).pop()
temptargets_input = params._.slice(6, 7).pop()
```

I also needed to remove:
```
model_input = params.model ? params._.slice(5, 6).pop() : false;
```

Comment (jasoncalabrese) [inline bin/oref0-get-profile.js]: Actually if the `if (params._.length > 6) {...}` was removed it works for me since only the `model` option was added

**Comment by jasoncalabrese** [comment]
Is the missing `-export-defaults` and `--update-preferences` flags the only thing know missing here?  I could probably do that tonight if @sulkaharo doesn't do it first.

I'm running this in production now, so far so good

File: bin/oref0-get-profile.js [hunk 4]
```diff
@@ -42,40 +42,36 @@ function updatePreferences (prefs) {
...
-    }
-
-    if (pumpsettings_input !== undefined && pumpsettings_input.indexOf('update-preferences') > 0) {
-      var prefs_input = process.argv.slice(3, 4).pop();
-	  if (!prefs_input) { console.error('usage: ', process.argv.slice(0, 2), '<preferences.json>'); process.exit(0); }
-      var prefs = require(cwd + '/' + prefs_input);
-      updatePreferences(prefs);
-      process.exit(0);
-    }
-
```
Comment (sulkaharo) [inline bin/oref0-get-profile.js]: This is the part that causes the export-defaults and update-preferences to be missing; the implementation is there but no way to invoke

**Comment by sulkaharo** [comment]
https://github.com/openaps/oref0/pull/163 has a new cleaner implementation that uses yargs

**Comment by sulkaharo** [comment]
That's a weird side-effect with yaml? Debugging

**Comment by scottleibrand** [comment]
Gonna go ahead and merge this.  We can merge the yargs fix directly once we have it ready.

**Pull request merged**